### PR TITLE
Test curl fixes

### PR DESCRIPTION
**Pull request by bewebi**
# Description

Update `testContainer.CurlShouldEventuallyRespond()` to use `cli.CurlFromPod()` to curl, which returns output from stderr and stdout separately
This ensures output is not mixed, as sometimes happens when `testContainer.Curl()` is used, which can lead to the response being parsed incorrectly

_Fill out any of the following sections that are relevant and remove the others_
## Test code changes
- `CurlFromPod()` returns `CurlResponse` object which has fields for `StdOut` and `StdErr`
- Add `WithCurlResponse()` transform which takes a `CurlResponse`, parses it, and returns an `*http.Response`
- `CurlEventuallyShouldOutput()`, which is already deprecated in favor of `CurlEventuallyShouldRespond()`, simply calls `CurlEventuallyShouldRespond()`
- `CurlEventuallyShouldRespond()` uses `cli.CurlFromPod()` in order to curl instead of `testContainer.Curl()` and uses `WithCurlResponse` transform, which handles the new response type and both HTTP 1.1 and 2
  - Add `buildCurlOptions()` helper function which factors out functionality from `buildCurlArgs()` that derives a `[]curl.Option` from `CurlOpts`
- Update `assertions.AssertEventualCurlResponse()` to account for new response type
  - Add intermediary `AssertEventualCurlReturnResponse()` function which returns the `*http.Response`
- Add `assertions.AssertCurlResponse()` and `assertions.AssertCurlReturnResponse()` which assert on curls outside of an `Eventually` block

# Context

solo-io/solo-projects#6337 was determined to be caused because the response from `testContainer.Curl()` would sometimes intersperse output from stdout and stderr, breaking assumptions we make when parsing the response for status code

For example we observed the following response string:
```
> GET /posts2 HTTP/1.1
> Host: ratelimit
> User-Agent: curl/7.58.0
> Accept: */*
> 
[
  {
    "userId": 1,
    "id": 1,
    "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    "body": "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
  },
<-- a whole lot of response -->
  {
    "userId": 9,
    "id": 89,
    "title": "sint soluta et vel magnam aut ut sed qui",
    "body": "repellat aut aperiam totam temporibus autem et\narchitecto magnam ut\nconsequatur qui cupiditate rerum quia soluta dignissimos nihil iure\ntempo< HTTP/1.1 200 OK
< date: Tue, 11 Jun 2024 15:45:59 GMT
< content-type: application/json; charset=utf-8
< report-to: {"group":"heroku-nel","max_age":3600,"endpoints":[{"url":"https://nel.heroku.com/reports?ts=1716219591&sid=e11707d5-02a7-43ef-b45e-2cf4d2036f7d&s=d%2FfZNhxR4feCi%2BeRiXTVGITxf22nbzi6OVvtC40LZJ8%3D"}]}
< reporting-endpoints: heroku-nel=https://nel.heroku.com/reports?ts=1716219591&sid=e11707d5-02a7-43ef-b45e-2cf4d2036f7d&s=d%2FfZNhxR4feCi%2BeRiXTVGITxf22nbzi6OVvtC40LZJ8%3D
< nel: {"report_to":"heroku-nel","max_age":3600,"success_fraction":0.005,"failure_fraction":0.05,"response_headers":["Via"]}
< x-powered-by: Express
< x-ratelimit-limit: 1000
<-- more headers -->
< content-length: 27520
< 
{ [27520 bytes data]
100 275re quas est"
  },
  {
    "userId": 9,
    "id": 90,
    "title": "ad iusto omnis odit dolor voluptatibus",
    "body": "minus omnis soluta quia\nqui sed adipisci voluptates illum ipsam voluptatem\neligendi officia ut in\neos soluta similique molestias praesentium blanditiis"
  },
<-- a bunch more of the response -->
  {
    "userId": 10,
    "id": 100,
    "title": "at nam consequatur ea labore ea harum",
    "body": "cupiditate quo est a modi nesciunt soluta\nipsa voluptas error itaque dicta in\nautem qui minus magnam et distinctio eum\naccusamus ratione error aut"
  }
]20  100 27520    0     0   248k      0 --:--:-- --:--:-- --:--:--  246k
* Connection #0 to host gateway-proxy left intact
```
Note that the `< HTTP/1.1 200 OK` substring is appended to a line of the response body, whereas we expect it to be on a new line

## Interesting decisions
 
We have `CurlEventuallyShouldOutput()` invoke `CurlEventuallyShouldResponsd()` rather than adapting it to use `cli.CurlFromPod()` both because the function is deprecated and not worth updating and because without additional effort to add async support for the `cli`, the resulting functionality would be identical to `CurlEventuallyShouldRespond()`
With so little time until GA release it doesn't make sense to throw a wrench into anything by removing the deprecated function entirely

This PR is based on changes originally included in #9600

## Testing steps

I ran the EE Upgrade tests from solo-io/solo-projects#6268 using this branch as the OSS dep and did not observe the flaky behavior we are hoping to address

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by bewebi** [comment]
/kick unclear why buildbot errored before starting

**Review by sheidkamp — DISMISSED** [review]

**Review by davidjumani — DISMISSED** [review]

File: pkg/utils/kubeutils/kubectl/cli.go [hunk 1]
```diff
@@ -44,6 +44,11 @@ func NewCli() *Cli {
 	}
 }
 
+type CurlResponse struct {
```
[the remark this answers is not in the document]
Comment (bewebi) [reply pkg/utils/kubeutils/kubectl/cli.go:49]: [74c9556](https://github.com/solo-io/gloo/pull/9606/commits/74c9556cdd95334dc073e345c390443314351171)

File: test/gomega/transforms/curl.go [hunk 2]
```diff
@@ -73,9 +106,19 @@ func processResponseHeader(line string) (string, string) {
 // Returns the status code if the line was processed, otherwise returns 0.
 func processResponseCode(line string) int {
 	// check for response status. the line with the response code will be in the format
-	// `< HTTP/1.1 <code> <message>`
-	if strings.HasPrefix(line, responseStatusPrefix) {
-		restOfLine := line[len(responseStatusPrefix):]
+	// `< HTTP/1.1 <code> <message>` or `< HTTP/2 <code> <message>`
```
[the remark this answers is not in the document]
Comment (bewebi) [reply test/gomega/transforms/curl.go:109]: [877d6d9](https://github.com/solo-io/gloo/pull/9606/commits/877d6d9d91cd79b7467927420ea29cdadb9217d9)

**Review by nfuden — APPROVED** [review]

**Review by sheidkamp — APPROVED** [review]

**Comment by bewebi** [comment]
bulldozer?

**Comment by bewebi** [comment]
bulldozer?

**Comment by bewebi** [comment]
bulldozer this is your last chance

**Pull request merged**